### DISCUSSION TITLE
Store Image with Disk Cache Locally

### DISCUSSION THREAD

**Discussion by IsaacTrevino** [coil-kt/coil#1508] [General]
Is it possible to store local images via Disk Cache? For example, If I have a URI or Bitmap, could I load the image using the request builder and retrieve the `diskCacheKey`? I have only gotten this to work with `memoryCacheKey` or when the DataSource is from an HTTP (e.g. URL).

My objective is to store images I have taken with the camera in the disk cache while offline. It appears that only the memory cache can only do that, whereas the disk cache only works with remote online sources.

My current implementation:
```typescript
class App: Application(), ImageLoaderFactory  {

    override fun newImageLoader(): ImageLoader {
        return ImageLoader.Builder(this)
            .diskCache {
                DiskCache.Builder()
                    .directory(this.cacheDir.resolve("image_cache"))
                    .maxSizePercent(1.00)
                    .build()
            }
            .logger(DebugLogger())
            .memoryCachePolicy(CachePolicy.DISABLED)
            .respectCacheHeaders(false)
            .build()
    }
}
```
Request: 
```typescript
// when offline
val request = ImageRequest
                    .Builder(context)
                    .data(uri) // or bitmap
                    .addHeader("Cache-Control", "no-cache")
                    .listener(
                        onError = { request: ImageRequest, result: ErrorResult ->
                            Log.i("$TAG+coil", result.toString())
                            Log.i("$TAG+coil", result.throwable.message.toString())
                        },
                        onSuccess = { request: ImageRequest, result: SuccessResult ->
                            Log.i("$TAG+coil", result.toString())
                            result.diskCacheKey?.let { it1 -> Log.i("$TAG+coil+disk", it1) }
                            result.memoryCacheKey?.let { it1 -> Log.i("$TAG+coil+mem", it1.key) }
                        }
                    )
                    .build()
                context.imageLoader.execute(request)
```

Perhaps it would be easier if I used the android native `ContentProvider` for my task, though disk caching seems to be the most ideal approach. And Glide seems to have complicated documentation. Thought I'd give Coil a try before looking for alternatives. 
https://developer.android.com/topic/performance/graphics/cache-bitmap#disk-cache

**Comment by irgaly** [comment]
I have faced same problem 😢 

I'd like to read a file from ContentProvider ( `content://...` path ), and cache it to DiskCache, then it should be used on next read because the `content://` url has no permission on later. 

* https://developer.android.com/training/data-storage/shared/photo-picker#persist-media-file-access
    * > By default, the system grants your app access to media files until the device is restarted or until your app stops. 

For now, Coil cannot used for this situation, because the ContentUriFetcher does not use DiskCache, while the NetworkFetcher uses DiskCache. So the DiskCache works only fetching images from networks.

https://github.com/coil-kt/coil/blob/29bc4189ce8daae1a82997a772e6963358654451/coil-core/src/androidMain/kotlin/coil3/fetch/ContentUriFetcher.kt

https://github.com/coil-kt/coil/blob/29bc4189ce8daae1a82997a772e6963358654451/coil-network-core/src/commonMain/kotlin/coil3/network/NetworkFetcher.kt#L65
